### PR TITLE
Refactor Game_Character classes + add move route test suite

### PR DESCRIPTION
**Pull request by fmatthew5876**
This is a large refactor for the `Game_Character` class heirarchy.

Depends on: https://github.com/EasyRPG/liblcf/pull/372

Goals
------

- [x] Remove use of `Main_Data:;game_data` from characters and map #1954
- [x] Restructure `Game_Character` to match RPG_RT design more closely. This will fix bugs we don't even know about yet, make it easier to fix all of the current bug set, and make it very easy to fix bugs discovered later.
- [x] Fix remaining incompatibilities with RPG_RT w.r.t characters and move routes
- [x] Simplify code and remove global variables

Issues Fixed
------------

- [x] Fix: #1645 
- [x] Fix: #2197 
- [x] Fix: #1859
- [x] Fix: #1852 
- [x] Fix: #2183
- [x] Fix: #2196 
- [x] Fix: #2207
- [x] Fix: #2224 
- [x] Fix: #2255 
- [x] Fix: #2286
- [x] Fix: #2316

Unit Test coverage:
- [ ] Fix: #1567 
- [ ] Fix: #1626 
- [ ] Fix: #1675 
- [ ] Fix: #1691 
- [ ] Fix: #1699 
- [x] Initialization
- [x] Movement / speed / frequency / stop count
- [x] Character Animation Types / frames
- [x] Vehicle animation frames
- [x] Move Routes
- [ ] Vehicle boarding / unboarding
- [ ] Vehicle ascent / descent
- [ ] Player movement when boarded syncs to vehicle
- [ ] Event Triggering
- [x] Flash
- [ ] Bush Depth
- [x] MoveTo()
- [ ] Player input and movement
- [ ] Player input and menu / encounter calling + situations which clear the flags
- [x] Panning
- [x] Save Counts
- [ ] Through flag properly set and reset in places
- [ ] MakeWay movement - maybe punt to later PR
- [ ] MakeWay jumping - maybe punt to later PR
- [ ] MakeWay land/disembark - maybe punt to later PR
- [ ] Event page refresh and page conditions

Additional Incompatibilities discovered and fixed
------------------

- [x] CommandHeroTranspareny resets the through flag
- [x] Verify stop animation flag is ignored for spin
- [x] through flag correctly used
- [x] Several additional cases when `menu_calling` and `encounter_calling` are cleared
- [x] airship will trigger collision events but not touch events. And only on `Game_Event`'s update sequence.
- [x] GetOnOffVehicle replaces diagonal directions with `sprite_direction`
- [x] Fix incorrect timing of spin with speed 2
- [ ] Post-boarding sets `phasing = hidden || slipThrough` ??

Testing Plan
-------------

- [x] Add a unit test suite
- [ ] Test a list of unique edge cases for this PR
- [ ] Retest the issue list
- [ ] A code reviewer should independently review this PR against RPG_RT in an RE tool. 
- [ ] Play a few games like a normal user would

Issue Retest List
----------

- [x] #1871 - retested a few cases
- [x] #1878 - tested Yume2kki music room
- [x] #1783  - tested a few multi-teleport cases
- [x] #1732
- [x] #1704 
- [x] #1707 - only move route chunk related
- [x] #1699 
- [x] #1696 
- [x] #1693 
- [x] #1691 
- [x] #1664 
- [x] #1659 
- [x] #1638 
- [x] #1604 
- [x] #1588 
- [x] #1936 
- [x] #1937 
- [x] #1938 
- [x] #579 
- [x] #937
- [x] #1267 
- [x] #1293 
- [x] #1346 
- [x] #1429 
- [x] #1585
- [x] #1587 
- [x] #1593 
- [x] #1654 
- [x] #1671

Lcf Chunk Renames to follow
---------
- [ ] move_route.repeated -> done
- [ ] sprite_direction -> facing
- [ ] through vs route_through -> something?
- [ ] sprite_transparent -> sprite_hidden

Dynrpg Info
-------------

Vtable and Structure layouts:
https://github.com/fmatthew5876/DynRPG/blob/master/Character.h

| Address | Class | Symbol |
| -- | -- | -- |
| 004C3D60 | Character | VTABLE |	
| 004A89A0 | Hero | VTABLE |
| 004B3BF4 | Vehicle | VTABLE |
| 004AAC8C | Event | VTABLE |
| 004A35D0 | SceneMap | Update() |
| 004AAF50 | SceneMap | RefreshAllPages() |
| 004B1938 | Interpreter | OnFinishEvent() |
| 004AEBD4 | Interpreter | CommandTeleport() |
| 004AEF30 | Interpreter | CommandSetEventLocation() |
| 004AFBF0 | Interpreter | CommandSetMoveRoute() |
| 004AF09C | Interpreter | CommandSwapEventLocation() |
| 004AD618 | Interpreter | CommandChangeParty() |
| 004AF610 | Interpreter | CommandPanScreen() |

Notes to self
-----------
- [x] Vehicles pending move on load / map change
- [x] Visible flag only used by hero, we use it in other places
- [x] CommandHeroTranspareny resets the through flag
- [x] Embarking a vehicle resets the through - test this?
- [x] Teleport / vehicle location / event location commands - review direction and facing
- [x] move route on inactive event
- [x] Check command halt all movement
- [x] Collision code sets menu calling
- [x] Get on/off vehicle not blocked by boarding or unboarding?
- [x] Check change party

### CODE REVIEW

**Comment by fmatthew5876** [comment]
| Address | Class | Name |
| -- | -- | -- |
| 004C1150 | Battler | ShouldMirrorBattleAnimations() - not used in this PR, but shows how battle aninimation mirroring depends on battler mirrored flag |
| ??? | SceneBattle | Logic which makes the character turn only if they perform an action on a single enemy - I need to find this again |
| 004977A4| SceneBattle | InitBattle() - initializes the battle all calls all the below routines |
| 004B4910 | SceneBattle | InitBattleActors() - does actor placement |
| 004BE8E0 | SceneBattle | InitBattleMonsters() - filters encounter condition and does enemy placement |
| 004B4B6C | SceneBattle | ResetPartyDirections() - Resets mirrored flag for all party at end of battle init sequence |
| 004BED2C | SceneBattle | ResetMonsterDirections() - Resets mirrored flag for all monster at end of battle init sequence |

This is rebased but the rebase was a little nasty. I need to retest again.

**Comment by Ghabry** [comment]
btw you can use ``REQUIRE`` with ``==, >, <, ...`` instead of ``REQUIRE_EQ/_GT``. It will still show the evaluated values when the test fails.

**Comment by fmatthew5876** [comment]
Interesting jump edge case:

```
Move UpRight
Jump
Move Forward
Land
Move Forward
```

So when a jump lands the `UpRight` direction gets replaced by Up

**Comment by fdelapena** [comment]
The PR fixes event collisions from airship, at least when calling battle, which is currently broken in 0.6.2.1 and master:

vehibug

Reported by @MudarraP

**Comment by fmatthew5876** [comment]
This is finally ready to be reviewed.

I still need to work through a lot of test cases but this is now in it's final state. I don't expect major changes from here, only bug fixes.

I didn't get to do all the full unit test coverage that I wanted to, but a significant portion is covered, including all the fundamental per frame behaviors of basic move routes.


This PR is essentially a rewrite of the entire `Game_Character` class hierarchy based on RE tools. The code is structured like RPG_RT so long term we will approach 100% correctness here very quickly.

It goes without saying that despite anyones best efforts, there will be regressions. Therefore I'd like us to take a hard think about how much is realistic via code review, how much testing each of us can realistically achieve on our own, and when it makes sense to just let it go to android testers so they can help us find bugs.

What I don't want to have happen is this sets for months. That is only going to make the situation worse as the code and my memory of it bitrots.

Know that once this goes in, I will be available and 100% committed to fixing regressions as they are found.

**Comment by Ghabry** [comment]
The compromise I can offer is that

- This is pushed to the Android BETA channel (currently up to 2500 testers)
- We ask for tests on Twitter etc.
- any PRs that conflicts with this will be delayed.

The review window will be around 4 weeks. This is so complex that it needs at least a weekend or two for proper reviewing & playtesting.

**Comment by Ghabry** [comment]
Regression:
Get Yume2kki 0.113a from https://yume2kki.fandom.com/wiki/Downloads

Load this save: 
[Save06.zip](https://github.com/EasyRPG/Player/files/5142931/Save06.zip)

You see miniteleporters (they activate on touch)
After walking on them the animation stops but in current master and RPG_RT they continue (looks like some animation type continuous page change bug)

**Comment by fmatthew5876** [comment]
> Regression:
> Get Yume2kki 0.113a from https://yume2kki.fandom.com/wiki/Downloads

Nice catch! Fixed

**Comment by Ghabry** [comment]
Patch for the emscripten test runner:

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index c4f9c88c..6b3747a7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -891,6 +891,10 @@ if(PLAYER_ENABLE_TESTS)
 	endif()
 	add_dependencies(check_player test_runner_player)
 	add_dependencies(check check_player)
+
+	if(CMAKE_SYSTEM_NAME STREQUAL "Emscripten")
+		set_property(TARGET test_runner_player APPEND_STRING PROPERTY LINK_FLAGS "-s ALLOW_MEMORY_GROWTH")
+	endif()
 endif()
 
 # Instrumentation framework
```

**Comment by Ghabry** [comment]
I still have problems with synchronizing the inputs but I can see that for the first 65 frames (this is move right, it desyncs on move up) that RPG_RT and your PR are exactly the same!

For comparison: Master desyncs on Frame 6.

So even this very basic test already shows improvements.

**Comment by Ghabry** [comment]
Regression:

[Map0003.zip](https://github.com/EasyRPG/Player/files/5146900/Map0003.zip)

On this map (uses 2k3 RTP) EV0006 faces right in your PR and down in 0.6.2.1 and RPG_RT.

**Comment by fmatthew5876** [comment]
> Regression:
> 
> [Map0003.zip](https://github.com/EasyRPG/Player/files/5146900/Map0003.zip)
> 
> On this map (uses 2k3 RTP) EV0006 faces right in your PR and down in 0.6.2.1 and RPG_RT. I think the sprite direction is always off-by-one. In your PR it is one higher than usual. Tested multiple directions.

Another simple one, rebased to master and fixed.

**Comment by Ghabry** [comment]
More accuracy stuff: The initialisation appears to be off-by-one:
I recorded now based on a hook after ``++frameCounter`` in RPG_RT and after ``IncFrame`` in Player.

[Map0002.zip](https://github.com/EasyRPG/Player/files/5147390/Map0002.zip)

A "Move right" is injected in Frame 50, which is processed in Frame 51. As you can see the stop count of the hero is correct after the move.

[frames.zip](https://github.com/EasyRPG/Player/files/5147405/frames.zip)

**Comment by fmatthew5876** [comment]
> I recorded now based on a hook after ++frameCounter in RPG_RT and after IncFrame in Player.

This is what is causing the difference. RPG_RT increments the frame counter *before* updating the scene, player does it at the end.


Detailed RPG_RT behavior:

* Increments the frame counter at the beginning of each frame
* SceneTitle: ++frame_counter
* When new game is selected:
* Calls a function which resets all game objects, including setting the frame counter to 0
* Sets up a teleport using the start position
* Calls the same `FinalizeTeleport` function used by event teleport, which runs pre-update (frame_counter == 0 when this happens)
* SceneMap: ++frame_count (==1)
* Runs the first map update


So essentially what happens here is the first time you hook RPG_RT, you only capture pre-update. When you hook player, you capture both pre-update and the first update.

So the fix here is to move the player frame increment before the first call to update.

**Comment by Ghabry** [comment]
this works almost flawless now. When I apply my Player patch below I get a perfectly synched input session (until a msg box opens because of subcommand and active flag).

the RPG_RT structure is

```
++frameCounter;
UpdateInput(); <- My hook (Input and Save) is here
switch (current_scene)->Update
```

In Player we will need something like this:

```
	if (update_scene) {
		IncFrame();
	}

	Input::Update(); <- Record/Inject input
	Audio().Update();

	if (update_scene) {
		/* Saved here */

		Scene::instance->Update();
	}
```

It would make more sense when I save after the update because then you get the state after the frame finished but the current solution only needs one hook so I prefer it (and in the end its the same, only the frame counter in the LSD differs by one).

**Comment by fmatthew5876** [comment]
Added `message_active`, so now you should be able to match on that one.

**Comment by Ghabry** [comment]
Together with my subcommand-hook I get now a perfect match.
Recording of anything on the map that is not depending on RNG should be possible now :)

**Comment by Ghabry** [comment]
Status update: 

I need a bit more time but I add a hard deadline here now:
The merge is in the last full September week (21-27).
Based on the last commits it is unlikely that I find a bug that needs a significant change so feel free to create new PRs that use this one as a base.

**Comment by Ghabry** [comment]
Could you rebase this?

I currently get:

```cpp
FAILED: CMakeFiles/check_player 
cd /data/tmp/easyrpg && /data/tmp/easyrpg/test_runner
easyrpg/easyrpg-player/tests/doctest.h:1114:5: runtime error: load of value 4294967295, which is not a valid value for type 'const lcf::rpg::EventPage::Trigger'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior easyrpg/easyrpg-player/tests/doctest.h:1114:5 in 
=================================================================
==28382==ERROR: AddressSanitizer: global-buffer-overflow on address 0x55a7b8ddc044 at pc 0x55a7b652d06b bp 0x7ffd578bb430 sp 0x7ffd578babe0
READ of size 1 at 0x55a7b8ddc044 thread T0
    #0 0x55a7b652d06a in __interceptor_strlen.part.0 (/data/tmp/easyrpg/test_runner+0x197906a)
    #1 0x7f5b912cd14e in std::char_traits<char>::length(char const*) /build/gcc/src/gcc-build/x86_64-pc-linux-gnu/libstdc++-v3/include/bits/char_traits.h:357:25
    #2 0x7f5b912cd14e in std::basic_ostream<char, std::char_traits<char> >& std::operator<<<std::char_traits<char> >(std::basic_ostream<char, std::char_traits<char> >&, char const*) /build/gcc/src/gcc-build/x86_64-pc-linux-gnu/libstdc++-v3/include/ostream:612:44
    #3 0x55a7b6e9818a in Game_Map::SetupCommon() easyrpg/easyrpg-player/src/game_map.cpp:304:6
    #4 0x55a7b6e960a5 in Game_Map::Setup(std::unique_ptr<lcf::rpg::Map, std::default_delete<lcf::rpg::Map> >) easyrpg/easyrpg-player/src/game_map.cpp:136:2
    #5 0x55a7b681dceb in MockGame::MockGame(MockMap) easyrpg/easyrpg-player/tests/mock_game.cpp:46:2
    #6 0x55a7b6799b3d in _DOCTEST_ANON_FUNC_48() easyrpg/easyrpg-player/tests/game_character_anim.cpp:260:17
    #7 0x55a7b6adc429 in doctest::Context::run() easyrpg/easyrpg-player/tests/doctest.h:5856:21
    #8 0x55a7b6ae27d9 in main easyrpg/easyrpg-player/tests/doctest.h:5940:71
    #9 0x7f5b90fc0151 in __libc_start_main (/usr/lib/libc.so.6+0x28151)
    #10 0x55a7b64e58ad in _start (/data/tmp/easyrpg/test_runner+0x19318ad)

0x55a7b8ddc044 is located 60 bytes to the left of global variable 'lcf::DBString::npos' defined in 'easyrpg/easyrpg-player/lib/liblcf/src/dbarray.cpp:16:41' (0x55a7b8ddc080) of size 4
0x55a7b8ddc044 is located 0 bytes to the right of global variable 'lcf::DBArrayAlloc::_empty_buf' defined in 'easyrpg/easyrpg-player/lib/liblcf/src/dbarray.cpp:15:49' (0x55a7b8ddc040) of size 4
SUMMARY: AddressSanitizer: global-buffer-overflow (/data/tmp/easyrpg/test_runner+0x197906a) in __interceptor_strlen.part.0
Shadow bytes around the buggy address:
```

**Comment by Ghabry** [comment]
@fmatthew5876  On some platforms the unit tests are crashing, can you check this?

File: src/game_character.h [hunk 1]
```diff
@@ -940,7 +940,7 @@ inline bool Game_Character::IsFacingLocked() const {
 }
 
 inline void Game_Character::SetFacingLocked(bool locked) {
-	data()->lock_facing = locked;
+	data()->lock_facing = locked || IsDirectionFixedAnimationType(GetAnimationType());
```
Comment (Ghabry) [inline src/game_character.h:971]: Minor: I find it confusing that SetFacingLocked can silently fail now. From such setters I would expect that they "work"

File: src/game_map.cpp [hunk 2]
```diff
@@ -149,8 +149,11 @@ void Game_Map::ResetPendingMove() {
 	}
 }
 
-void Game_Map::Setup() {
+void Game_Map::Setup(std::unique_ptr<RPG::Map> map_in) {
```
Comment (Ghabry) [inline src/game_map.cpp]: uh, this is nice, this will allow implementing a ``--map`` (without ID) feature

File: src/game_event.cpp [hunk 3]
```diff
@@ -55,6 +55,20 @@ void Game_Event::SetSaveData(lcf::rpg::SaveMapEvent save)
 	data()->ID = event->ID;
 	SetMapId(map_id);
 
+	// Sanity checks on bad save data
+	auto checkMoveRoute = [&](const auto& mr, auto& idx, const auto& name) {
+		const auto n = static_cast<int>(mr.move_commands.size());
+		if (idx < 0 || idx > n) {
+			idx = n;
+			Output::Warning("Event {}: Save Data invalid {}={}. Fixing ...", data()->ID, name, idx);
```
Comment (Ghabry) [inline src/game_event.cpp:63]: The warning should mention that it is a MoveRoute

File: src/game_character.cpp [hunk 4]
```diff
@@ -216,6 +216,9 @@ void Game_Character::UpdateMoveRoute(int32_t& current_index, const lcf::rpg::Mov
 	if (is_overwrite && !IsMoveRouteOverwritten()) {
 		return;
 	}
+	const auto num_commands = static_cast<int>(current_route.move_commands.size());
+	assert(current_index >= 0);
+	assert(current_index <= num_commands);
```
Comment (Ghabry) [inline src/game_character.cpp:220]: Can this trigger when a hacked LDB is used? If yes this should be a warning and not an assert

**Review by Ghabry — COMMENTED** [review]
-

File: src/game_event.cpp (hunk 3 above)
Comment (fmatthew5876) [reply src/game_event.cpp:63]: The lambda gets passed in `move_route_index` or `original_move_route_index` string name

File: src/game_character.cpp (hunk 4 above)
Comment (fmatthew5876) [reply src/game_character.cpp:220]: No, but it can trigger on a corrupted save game. I found this from an unrelated bug, that if the move route index is invalid it causes an infinite loop. The piece of code you previously commented on sanitizes the save data to prevent this from ever happening, and so the assert here is used to make sure that invariant is held.

I will add a comment here explaining this

File: src/game_character.h (hunk 1 above)
Comment (fmatthew5876) [reply src/game_character.h:977]: This is how RPG_RT does it. To make the interface more sane we would need a careful refactor to ensure the correct behavior is not disturbed.

I say punt for now.

**Review by Ghabry — APPROVED** [review]
Let's get the counter close to 200
